timeseries/src/autogluon/timeseries/models/chronos/model.py

```diff
@@ -0,0 +1,366 @@
+import pandas as pd
+from transformers import AutoConfig, AutoModelForCausalLM, AutoModelForSeq2SeqLM
+
+from autogluon.common.loaders import load_pkl
+from autogluon.timeseries.dataset.ts_dataframe import ITEMID, TIMESTAMP, TimeSeriesDataFrame
+from autogluon.timeseries.models.abstract import AbstractTimeSeriesModel
+from autogluon.timeseries.utils.forecast import get_forecast_horizon_index_ts_dataframe
+
+from .chronos import ChronosConfig, ChronosPipeline, ChronosPretrainedModel
+
+# NOTE: We avoid imports for torch and lightning.pytorch at the top level and hide them inside class methods.
+# This is done to skip these imports during multiprocessing (which may cause bugs)
+
+logger = logging.getLogger(__name__)
+
+
+# allowed HuggingFace model paths with custom parameter definitions
+MODEL_CONFIGS = {
+    "amazon/chronos-t5-tiny": {
+        "num_gpus": 0,  # minimum number of required GPUs
+    },
+    "amazon/chronos-t5-mini": {"num_gpus": 0},
+    "amazon/chronos-t5-small": {"num_gpus": 1},
+    "amazon/chronos-t5-base": {"num_gpus": 1},
+    "amazon/chronos-t5-large": {"num_gpus": 1},
+}
+
+
+class ChronosInferenceDataset:
+    """A container for time series datasets that implements the ``torch.utils.data.Dataset`` interface"""
+
+    def __init__(
+        self,
+        target_df: TimeSeriesDataFrame,
+        context_length: int,
+        target_column: str = "target",
+    ):
+        assert target_df is not None
+        assert target_df.freq, "Initializing GluonTS data sets without freq is not allowed"
+
+        self.target_column = target_column
+        self.context_length = context_length
+        self.target_array = target_df[target_column].to_numpy(dtype=np.float32)
+        self.freq = target_df.freq
+        self._set_indptr(target_df=target_df)
+
+    def _set_indptr(self, target_df: TimeSeriesDataFrame):
+        """Replace inefficient groupby ITEMID with indptr that stores start:end of each time series"""
+
+        # TODO: refactor into generic iterator
+
+        item_id_index = target_df.index.get_level_values(ITEMID)
+        indices_sizes = item_id_index.value_counts(sort=False)
+        self.item_ids = indices_sizes.index  # shape [num_items]
+        cum_sizes = indices_sizes.values.cumsum()
+        self.indptr = np.append(0, cum_sizes).astype(np.int32)
+        self.start_timestamps = target_df.reset_index(TIMESTAMP).groupby(level=ITEMID, sort=False).first()[TIMESTAMP]
+        assert len(self.item_ids) == len(self.start_timestamps)
+
+    def __len__(self):
+        return len(self.indptr) - 1  # noqa
+
+    def _get_context(self, a: np.ndarray, pad_value=np.nan):
+        assert self.context_length > 0
+        a = a[-self.context_length :]
+        pad_size = self.context_length - len(a)
+        if pad_size > 0:
+            pad = np.full(shape=(pad_size,), fill_value=pad_value)
+            a = np.concatenate((pad, a))
+        return a
+
+    def get_full_item(self, idx) -> Dict[str, Any]:
```

Comment:
This method seems to be unused

Reply:
Here for completeness. We will likely have to extend this when covariates come into play.